### PR TITLE
Check node upgrade before running e2e_test.go

### PR DESCRIPTION
**Pull request by hieuvubk**
Closes: #4066 

## What is the purpose of the change

Check that nodes are running successfully after `runProposalUpgrade`

## Brief Changelog

Create func `NodeStatus` for get status of node, if there are any problems return err
Check all nodes status at the end of func `RunUpgrade`

## Testing and Verifying

*(Please pick one of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added unit test that validates ...*
  - *Added integration tests for end-to-end deployment with ...*
  - *Extended integration test for ...*
  - *Manually verified the change by ...*

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: tests/e2e/configurer/chain/node.go
Comment (p0mvn) [inline tests/e2e/configurer/chain/node.go:167]: Does it make sense to return this status? Otherwise, why would we parse it? Maybe there is no use case right now but there might be in the future

File: tests/e2e/configurer/chain/node.go
Comment (p0mvn) [inline tests/e2e/configurer/chain/node.go:151]: nit: suggest putting this in `commands.go` and naming just `Status`

File: tests/e2e/configurer/upgrade.go
Comment (p0mvn) [inline tests/e2e/configurer/upgrade.go:168]:
```suggestion
				uc.t.Errorf("Node is not running after upgrade, chain-id %s, node %s", chainConfig.Id, node.Name)
```

File: tests/e2e/configurer/upgrade.go
Comment (p0mvn) [inline tests/e2e/configurer/upgrade.go:171]: If there is no error, can we add a log that a node with id X and name Y has upgraded successfully please

**Review by p0mvn — COMMENTED** [review]
Awesome. LGTM once comments are addressed!

File: tests/e2e/configurer/chain/node.go
Comment (hieuvubk) [reply tests/e2e/configurer/chain/node.go:167]: I found it in the SDK, it would be better if we unmarshal the response according to its struct.

File: tests/e2e/configurer/upgrade.go
Comment (stackman27) [reply tests/e2e/configurer/upgrade.go:172]: ^^ Also following up on this, is it possible to add "good things to know" in e2e readme where you show this failure happening and ways to fix it?

File: tests/e2e/configurer/chain/commands.go
Comment (p0mvn) [inline tests/e2e/configurer/chain/commands.go:333]: What I meant was returning the `result` from this function.

File: tests/e2e/configurer/chain/node.go
Comment (p0mvn) [reply tests/e2e/configurer/chain/node.go]: https://github.com/osmosis-labs/osmosis/pull/4150#discussion_r1091427246

File: tests/e2e/configurer/upgrade.go
Comment (p0mvn) [inline tests/e2e/configurer/upgrade.go:168]:
```suggestion
				uc.t.Errorf("node is not running after upgrade, chain-id %s, node %s", chainConfig.Id, node.Name)
```

**Comment by p0mvn** [comment]
Hi @hieuvubk . Left some more comments, also e2e seems to be failing. I tried rerunning a few times

File: tests/e2e/configurer/chain/commands.go
Comment (p0mvn) [inline tests/e2e/configurer/chain/commands.go:336]:
```suggestion
	return result, nil
```

**Review by p0mvn — COMMENTED** [review]
lgtm once CI is passing

File: tests/e2e/configurer/upgrade.go
Comment (mattverse) [inline tests/e2e/configurer/upgrade.go:166]: just out of curiousity, do we not need to check on the status returned but only do err checking to determine if the node is running / or not?

Comment (p0mvn) [reply tests/e2e/configurer/upgrade.go:166]: If we get a correct response from `osmosisd status` and it's in the desired format (can be parsed), we can conclude that the node is running.

I don't think any of the result fields are useful for our purposes since we just want to make sure that a node is running

Comment (hieuvubk) [reply tests/e2e/configurer/upgrade.go:166]: From the last check, seem that we cant unmarshal the `Status` response to the original struct cause it's already converted to string obj.
So I think just check the cmd response & no return err is enough. cc @RusAkh

Comment (hieuvubk) [reply tests/e2e/configurer/upgrade.go:166]: And use `any` type to unmarshal

File: tests/e2e/configurer/chain/commands.go
Comment (p0mvn) [inline tests/e2e/configurer/chain/commands.go:318]: I'm not sure if this parsing right. What is the final structure / format of the `result`? How can we be sure that it parses correctly?

Comment (hieuvubk) [reply tests/e2e/configurer/chain/commands.go:318]: its a json stringify

Comment (hieuvubk) [reply tests/e2e/configurer/chain/commands.go:318]: Got some problems when converting `string` to golang native types

Comment (hieuvubk) [reply tests/e2e/configurer/chain/commands.go:318]: Got some problems when converting `string` to golang native types

Comment (p0mvn) [reply tests/e2e/configurer/chain/commands.go:318]: What does the problem / error say?

Comment (p0mvn) [reply tests/e2e/configurer/chain/commands.go:318]: At what commit exactly were you seeing this error?

Comment (hieuvubk) [reply tests/e2e/configurer/chain/commands.go:318]: https://github.com/osmosis-labs/osmosis/pull/4150/commits/af9decbe9171a3f69407878bb89c053a5e6d9e0d

Comment (hieuvubk) [reply tests/e2e/configurer/chain/commands.go]: its fixed, sr for take it long time

**Review by p0mvn — APPROVED** [review]
LGTM, great work

File: tests/e2e/configurer/upgrade.go
Comment (p0mvn) [reply tests/e2e/configurer/upgrade.go:172]: @hieuvubk can you please address this suggestion from stackman27 before we merge?

**Comment by p0mvn** [comment]
@hieuvubk please take a look at this comment before merging: https://github.com/osmosis-labs/osmosis/pull/4150/files#r1095954616

File: tests/e2e/configurer/upgrade.go
Comment (hieuvubk) [reply tests/e2e/configurer/upgrade.go:172]: yes sure I just missed this cmt thanks u @stackman27

**Review by p0mvn — APPROVED** [review]
Awesome. Thanks!

**Pull request merged**
